### PR TITLE
Core/Battleground: Areatriggers will now check if player leaves the start area before the battleground begins in the following battlegrounds:

### PR DESCRIPTION
**Pull request by Golrag**
- Arathi Bassin
- Alterac Valley
- Eye of the Storm
- Isle of Conquest
- Warsong Gulch

https://github.com/TrinityCore/TrinityCore/pull/15841

### CODE REVIEW

**Comment by soulfrost** [comment]
there's no reason in this code of old checks still in core

**Comment by Shauren** [comment]
@soulfrost go away. This is a good PR.

**Comment by soulfrost** [comment]
there's the reason of this pr if this allready handled in other place? lol

**Comment by tkrokli** [comment]
@soulfrost : you are misunderstanding the PR title. It is **because of** how areatriggers are working now in the rest of the source that the changes in this PR are needed.

**Comment by Golrag** [comment]
People getting caught by the _CheckSafePosition are avoiding the areatriggers, so you have the possibility to log it, and say they cheated (if the battleground has the areatrigger script)

**Comment by soulfrost** [comment]
better reimplent method to check other thing: if player is in AT phase and range but server dont get cmsg_at -> then ban/kick player

**Comment by Kinzcool** [comment]
Using the starting position isn't the right way of doing this; you have to use the coordinates of the PortLocIDs, that is the coordinates of the areatriggers in WorldSafeLoc.dbc

**Comment by Golrag** [comment]
Any idea  how I can get them or should I look them up manually ?

**Comment by joschiwald** [comment]
there is a table `areatrigger_teleport`

**Comment by Golrag** [comment]
Something like this ? For Arathi Bassin
```cpp
case 6635: // Horde Start
case 6634: // Alliance Start
    if (GetStatus() == STATUS_WAIT_JOIN && !entered)
    {
        uint32 startWorldSafeLocs[] = { 3706, 3705 };
        if (WorldSafeLocsEntry const* start = sWorldSafeLocsStore.LookupEntry(startWorldSafeLocs[Battleground::GetTeamIndexByTeamId(player->GetBGTeam())]))
            player->TeleportTo(start->MapID, start->Loc.X, start->Loc.Y, start->Loc.Z, start->Facing);
        else
        {
            TC_LOG_ERROR("server.error", "Could not find WorldSafeLocation: %u", startWorldSafeLocs[Battleground::GetTeamIndexByTeamId(player->GetBGTeam())]);
            Position const* startPos = GetTeamStartPosition(Battleground::GetTeamIndexByTeamId(player->GetBGTeam()));
            player->TeleportTo(GetMapId(), startPos->GetPositionX(), startPos->GetPositionY(), startPos->GetPositionZ(), startPos->GetOrientation());
        }
    }
break;
```

Note: I got these WorldSafeLocs: http://puu.sh/lJjOa/eb44be3ca1.png

**Comment by Golrag** [comment]
update to https://github.com/TrinityCore/TrinityCore/commit/cdb5bd3eb78ea40fbd927c5662e1dad778bf12d5
Also fixed Arathi Bassin

File: src/server/game/Battlegrounds/Zones/BattlegroundAV.cpp [hunk 1]
```diff
@@ -493,11 +493,16 @@ void BattlegroundAV::RemovePlayer(Player* player, ObjectGuid /*guid*/, uint32 /*
...
 void BattlegroundAV::HandleAreaTrigger(Player* player, uint32 trigger, bool entered)
 {
-    if (GetStatus() != STATUS_IN_PROGRESS)
-        return;
-
     switch (trigger)
     {
+        case 6633: // Horde Start
+        case 6632: // Alliance Start
+            if (GetStatus() == STATUS_WAIT_JOIN && entered)
```
Comment (Shauren) [inline src/server/game/Battlegrounds/Zones/BattlegroundAV.cpp]: is entered == true correct here?

Comment (Golrag) [inline src/server/game/Battlegrounds/Zones/BattlegroundAV.cpp]: Yes, this areatrigger is right infront of the gate (iirc the position of the gate on retail is different than on trinitycore)

Comment (Shauren) [inline src/server/game/Battlegrounds/Zones/BattlegroundAV.cpp]: ok, was just wondering if this is intended or a typo

**Comment by Aokromes** [comment]
@Golrag can you update this PR plz?

**Comment by Golrag** [comment]
Yea sure, just to know: should I work with the WorldStates in this PR or do you guys prefer to continue with the start position ?

**Comment by joschiwald** [comment]
@Golrag why you are not using teleports from `areatrigger_teleport` table?

**Comment by Shauren** [comment]
@joschiwald because areatrigger_teleport will teleport always, not only if the battleground isn't started
@Golrag yes, WorldSafeLocs would be preferred

File: src/server/game/Battlegrounds/Battleground.cpp [hunk 2]
```diff
@@ -1076,6 +1076,12 @@ void Battleground::StartBattleground()
         TC_LOG_DEBUG("bg.arena", "Arena match type: %u for Team1Id: %u - Team2Id: %u started.", m_ArenaType, m_ArenaTeamIds[TEAM_ALLIANCE], m_ArenaTeamIds[TEAM_HORDE]);
 }
 
+void Battleground::TeleportPlayerToExploitLocation(Player* player)
+{
+    if (WorldSafeLocsEntry const* loc = GetExploitTeleportLocation(player->GetTeamId()))
```
Comment (Shauren) [inline src/server/game/Battlegrounds/Battleground.cpp]: one thing that just came to my mind - shouldn't GetBGTeam be used here if this detection also applies to arenas (are there worldsafelocs for that?)

Comment (Golrag) [inline src/server/game/Battlegrounds/Battleground.cpp]: Haven't looked at arenas tbh (Can't queue for an arena so I wouldn't be able to test it right now). I find the whole team thing very confusing. Which enum matches with GetBGTeam() ?

Comment (Shauren) [inline src/server/game/Battlegrounds/Battleground.cpp]: enum Team

**Comment by joschiwald** [comment]
let HandleAreaTrigger return a bool to allow or disallow port like in opvp classes, to solve the "always teleport thing"

**Pull request merged**